chips/nrf51/src/ble_advertising_driver.rs

```diff
@@ -0,0 +1,272 @@
+                self.busy.set(false);
+                ReturnCode::SUCCESS
+            }
+            (2, false) => self.radio.set_adv_txpower(data),
+            (3, false) => {
+                if data < 20 {
+                    self.advertisement_interval.set(20 * <A::Frequency>::frequency() / 1000);
+                } else if data > 10240 {
+                    self.advertisement_interval.set(10240 * <A::Frequency>::frequency() / 1000);
+                } else {
+                    self.advertisement_interval
+                        .set((data as u32) * <A::Frequency>::frequency() / 1000);
+                }
+                ReturnCode::SUCCESS
+            }
+            (4, false) => {
+                self.offset.set(0);
+                self.radio.clear_adv_data();
+                ReturnCode::SUCCESS
+            }
+            (_, true) => ReturnCode::EBUSY,
+            (_, _) => ReturnCode::ENOSUPPORT,
+        }
+    }
+
+    fn allow(&self, appid: AppId, allow_num: usize, slice: AppSlice<Shared, u8>) -> ReturnCode {
+        match (allow_num, self.busy.get()) {
+            // See this as a giant case switch or if else statements
+            (e @ BLE_HS_ADV_TYPE_FLAGS, false) |
+            (e @ BLE_HS_ADV_TYPE_INCOMP_UUIDS16, false) |
+            (e @ BLE_HS_ADV_TYPE_COMP_UUIDS16, false) |
+            (e @ BLE_HS_ADV_TYPE_INCOMP_UUIDS32, false) |
+            (e @ BLE_HS_ADV_TYPE_COMP_UUIDS32, false) |
+            (e @ BLE_HS_ADV_TYPE_INCOMP_UUIDS128, false) |
+            (e @ BLE_HS_ADV_TYPE_COMP_UUIDS128, false) |
+            (e @ BLE_HS_ADV_TYPE_INCOMP_NAME, false) |
+            (e @ BLE_HS_ADV_TYPE_COMP_NAME, false) |
+            (e @ BLE_HS_ADV_TYPE_TX_PWR_LVL, false) |
+            (e @ BLE_HS_ADV_TYPE_SLAVE_ITVL_RANGE, false) |
+            (e @ BLE_HS_ADV_TYPE_SOL_UUIDS16, false) |
+            (e @ BLE_HS_ADV_TYPE_SOL_UUIDS128, false) |
+            (e @ BLE_HS_ADV_TYPE_SVC_DATA_UUID16, false) |
+            (e @ BLE_HS_ADV_TYPE_PUBLIC_TGT_ADDR, false) |
+            (e @ BLE_HS_ADV_TYPE_RANDOM_TGT_ADDR, false) |
+            (e @ BLE_HS_ADV_TYPE_APPEARANCE, false) |
+            (e @ BLE_HS_ADV_TYPE_ADV_ITVL, false) |
+            (e @ BLE_HS_ADV_TYPE_SVC_DATA_UUID32, false) |
+            (e @ BLE_HS_ADV_TYPE_SVC_DATA_UUID128, false) |
+            (e @ BLE_HS_ADV_TYPE_URI, false) |
+            (e @ BLE_HS_ADV_TYPE_MFG_DATA, false) => {
+                let ret = self.app
+                    .enter(appid, |app, _| {
+                        app.app_write = Some(slice);
+                        ReturnCode::SUCCESS
+                    })
+                    .unwrap_or_else(|err| match err {
+                        Error::OutOfMemory => ReturnCode::ENOMEM,
+                        Error::AddressOutOfBounds => ReturnCode::EINVAL,
+                        Error::NoSuchApp => ReturnCode::EINVAL,
+                    });
+                if ret == ReturnCode::SUCCESS {
+                    self.set_adv_data(e)
```

Comment:
Why do all this `e @ ...` nonsense above instead of just calling `self.set_adv_data(allow_num)` here?